test/example.c

```diff
@@ -19,13 +19,94 @@
 #  define TESTFILE "foo.gz"
 #endif
 
-#define CHECK_ERR(err, msg) { \
-    if (err != Z_OK) { \
-        fprintf(stderr, "%s error: %d\n", msg, err); \
-        exit(1); \
+#define SUCCESSFUL                 1
+#define FAILED_WITH_ERROR_CODE     0
+#define FAILED_WITHOUT_ERROR_CODE -1
+
+typedef struct test_result_s {
+    int           result; /* One of: SUCCESSFUL,
+                          FAILED_WITH_ERROR_CODE, or
+                          FAILED_WITHOUT_ERROR_CODE*/
+    int           error_code; /* error code if success is FAILED_WITH_ERROR_CODE */
+    int           line_number;
+    z_const char* message;
+    z_const char* extended_message;
+} test_result;
+
+#define STRING_BUFFER_SIZE 100
+char string_buffer[STRING_BUFFER_SIZE];
+
+#define RETURN_ON_ERROR_WITH_MESSAGE(_error_code, _message, _result) { \
```

Comment:
Couldn't this declare a local test_result and thereby keep the same name and interface as CHECK_ERR?

Reply:
Sort of. We could create a new scope block in the macro.

C89 requires variables be declared at the top of the scope block. Without creating a new scope block, I needed to pass in the already-created variable. But a new scope block & local test_result makes sense.

That said, it would be a significant-enough difference from CHECK_ERR I think. CHECK_ERR would always exit(1) on error. This will always return. So maybe a different name should still be used? Especially since there are other macros for the other cases.

What do you think?

Reply:
I'm a fan of keeping the diff size small.  It's kosher IMHO to redefine the meaning of CHECK_ERR slightly; the name never did actually promise to abort.

Check out the branch I sent, I think it's tider...

Reply:
I looked at the branch.
I don't mind repurposing CHECK_ERR. However, it looks odd to mix CHECK_ERR and RETURN_WITH_MESSAGE / RETURN_WITH_EXTENDED_MESSAGE.

Maybe those could be renamed to SKIP_ERR_CHECK / CHECK_ERR_WITH_EXTENDED_MESSAGE or the likes?

Also, your branch needs to introduce a scope block to remain compatible with C89.

Reply:
Now you're talkin'.  I like your new suggested names.

gcc -std=c89 seems to pass on my branch...?

Reply:
I meant keeping the CHECK_ERR name and replacing RETURN_WITH_MESSAGE with SKIP_ERR_CHECK and replacing RETURN_WITH_EXTENDED_MESSAGE (now gone) with CHECK_ERR_WITH_EXTENDED_MESSAGE.

Now that RETURN_WITH_EXTENDED_MESSAGE is gone, maybe we can go with just CHECK_ERR and DONT_CHECK_ERR?



The scopes you are missing are in the macros
RETURN_SUCCESS
RETURN_SUCCESS_WITH_MESSAGE
RETURN_SUCCESS_WITH_EXTENDED_MESSAGE

It looks like -std=c89 or -ansi only disable certain things, which doesn't include the variables-at-top-of-scope-block requirement:
https://gcc.gnu.org/onlinedocs/gcc-4.2.4/gcc/C-Dialect-Options.html
It looks like -pedantic *may* hit it ("it finds some non-ISO practices, but not all"):
https://gcc.gnu.org/onlinedocs/gcc-4.2.4/gcc/Warning-Options.html#Warning-Options

Reply:
(I meant -std=c89 and -pedantic may hit it. Not -pedantic alone.)

Reply:
Thanks, adding -pedantic did find one in test_large_inflate, but the three macros you're looking at do declare scopes, look again :-)

I kind of like the RETURN_WITH_MESSAGE name for something that just returns with a message and doesn't check... although maybe it should be RETURN_FAILURE_WITH_MESSAGE (vs RETURN_SUCCESS_WITH_MESSAGE)?

Reply:
Oh wait. Right.
I blanked on macros being weird and thought the outer {} didn't count as scope. I'll update this CL to remove the extra {}s I added which we don't need.

That said, I'm not sure I see the issue in test_large_inflate.

Reply:
I like your idea of RETURN_FAILURE_WITH_MESSAGE and *SUCCESS*.
In fact, we can leave out the _WITH_MESSAGE. Now I think we're landing on some good names. Glad you pushed for CHECK_ERR :D

I uploaded a commit with CHECK_ERR, and RETURN_SUCCESS, RETURN_FAILURE.